### PR TITLE
 Integrate TSDF components into SLAM pipeline.

### PR DESCRIPTION
**Pull request by kdaun**
[RFC=0019](https://github.com/googlecartographer/rfcs/blob/master/text/0019-probability-grid-and-submap2d-restructuring.md)

### CODE REVIEW

File: cartographer/mapping/internal/2d/scan_matching/ceres_scan_matcher_2d.cc [hunk 1]
```diff
@@ -70,12 +71,22 @@ void CeresScanMatcher2D::Match(const Eigen::Vector2d& target_translation,
                                    initial_pose_estimate.rotation().angle()};
   ceres::Problem problem;
   CHECK_GT(options_.occupied_space_weight(), 0.);
-  problem.AddResidualBlock(
-      CreateOccupiedSpaceCostFunction2D(
-          options_.occupied_space_weight() /
-              std::sqrt(static_cast<double>(point_cloud.size())),
-          point_cloud, grid),
-      nullptr /* loss function */, ceres_pose_estimate);
+  CHECK(grid.grid_type() != GridType::NONE);
```
Comment (gaschler) [inline cartographer/mapping/internal/2d/scan_matching/ceres_scan_matcher_2d.cc]: As discussed, get rid of GridType::None.
1. make Grid2D::grid_type() const pure virtual
2. drop Grid2D::grid_type_, return const values instead

File: cartographer/mapping/internal/2d/scan_matching/ceres_scan_matcher_2d.cc [hunk 2]
```diff
@@ -70,12 +71,22 @@ void CeresScanMatcher2D::Match(const Eigen::Vector2d& target_translation,
...
+            point_cloud, grid),
+        nullptr /* loss function */, ceres_pose_estimate);
+  } else if (grid.grid_type() == GridType::TSDF) {
+    problem.AddResidualBlock(
+        CreateTSDFMatchCostFunction2D(
+            options_.occupied_space_weight() /
+                std::sqrt(static_cast<double>(point_cloud.size())),
+            point_cloud, static_cast<const TSDF2D&>(grid)),
+        nullptr /* loss function */, ceres_pose_estimate);
+  }
```
Comment (gaschler) [inline cartographer/mapping/internal/2d/scan_matching/ceres_scan_matcher_2d.cc]: else log fatal

File: cartographer/mapping/map_builder_test.cc [hunk 3]
```diff
@@ -15,6 +15,7 @@
  */
 
 #include "cartographer/mapping/map_builder.h"
+#include <mapping/2d/grid_2d.h>
```
Comment (gaschler) [inline cartographer/mapping/map_builder_test.cc]: style: full path, "", sort below

File: cartographer/mapping/map_builder_test.cc [hunk 4]
```diff
@@ -111,21 +132,27 @@ class MapBuilderTest : public ::testing::Test {
...
+    MapBuilderTestParamterizedByGridTypeAndDimensions,
+    MapBuilderTestParamterizedByGridTypeAndDimensions,
+    ::testing::Values(std::make_pair(GridType::PROBABILITY_GRID, 2),
+                      std::make_pair(GridType::PROBABILITY_GRID, 3),
+                      std::make_pair(GridType::TSDF, 2)));
 
-TEST_F(MapBuilderTest, TrajectoryAddFinish3D) {
-  SetOptionsTo3D();
+TEST_P(MapBuilderTestParamterizedByGridTypeAndDimensions, TrajectoryAddFinish) {
+  if (GetParam().second == 3) SetOptionsTo3D();
```
Comment (gaschler) [inline cartographer/mapping/map_builder_test.cc]: check if you can move these to the constructors of MapBuilderTestParamterizedByGridType and so on,
also for client_server_test if this works

File: cartographer/cloud/internal/client_server_test.cc [hunk 5]
```diff
@@ -178,6 +179,22 @@ class ClientServerTest : public ::testing::Test {
...
+        ->set_range_data_inserter_type(
+            ::cartographer::mapping::proto::RangeDataInserterOptions::
+                TSDF_INSERTER_2D);
+    trajectory_builder_options_.mutable_trajectory_builder_2d_options()
+        ->mutable_submaps_options()
+        ->mutable_grid_options_2d()
+        ->set_grid_type(::cartographer::mapping::proto::GridOptions2D::TSDF);
+    trajectory_builder_options_.mutable_trajectory_builder_2d_options()
+        ->mutable_ceres_scan_matcher_options()
+        ->set_occupied_space_weight(10.0);
```
Comment (gaschler) [inline cartographer/cloud/internal/client_server_test.cc]: this differs from map_builder_test, intended?

File: cartographer/cloud/internal/client_server_test.cc [hunk 6]
```diff
@@ -212,13 +229,27 @@ class ClientServerTest : public ::testing::Test {
   int number_of_insertion_results_;
 };
 
+class ClientServerTest : public ClientServerTestBase<::testing::Test> {};
+class ClientServerTestParamterizedByGridType
```
Comment (gaschler) [inline cartographer/cloud/internal/client_server_test.cc]: spelling: Parameterized (or just drop Parameterized)

File: cartographer/mapping/map_builder_test.cc (hunk 4 above)
Comment (kdaun) [reply cartographer/mapping/map_builder_test.cc]: MapBuilderTestParamterizedByGridType ist a instantiation of the MapBuilderTestBase template, which is used for parameterized and non-parameterized tests. Therefore, the functions cannot be moved to the constructor.

File: configuration_files/trajectory_builder_2d.lua [hunk 7]
```diff
@@ -82,10 +82,10 @@ TRAJECTORY_BUILDER_2D = {
           num_normal_samples = 4,
           sample_radius = 0.5,
         },
-        project_sdf_distance_to_scan_normal = false,
+        project_sdf_distance_to_scan_normal = true,
         update_weight_range_exponent = 0,
-        update_weight_angle_scan_normal_to_ray_kernel_bandwith = 0,
-        update_weight_distance_cell_to_hit_kernel_bandwith = 0,
+        update_weight_angle_scan_normal_to_ray_kernel_bandwith = 0.5,
```
[the remark this answers is not in the document]
Comment (kdaun) [reply configuration_files/trajectory_builder_2d.lua]: Before, these numbers were initialized to valid but not meaningful values. In order for the tests to work we need meaningful values.

File: cartographer/mapping/2d/submap_2d.h [hunk 8]
```diff
@@ -47,7 +47,7 @@ class Submap2D : public Submap {
   explicit Submap2D(const proto::Submap2D& proto,
                     ValueConversionTables* conversion_tables);
 
-  proto::Submap ToProto(bool include_probability_grid_data) const override;
+  proto::Submap ToProto(bool include_grid_data) const override;
```
[the remark this answers is not in the document]
Comment (kdaun) [reply cartographer/mapping/2d/submap_2d.h]: Yes, done.

File: cartographer/cloud/internal/client_server_test.cc (hunk 5 above)
Comment (kdaun) [reply cartographer/cloud/internal/client_server_test.cc]: It's unified now.

File: cartographer/mapping/2d/probability_grid.h [hunk 9]
```diff
@@ -50,6 +50,8 @@ class ProbabilityGrid : public Grid2D {
   bool ApplyLookupTable(const Eigen::Array2i& cell_index,
                         const std::vector<uint16>& table);
 
+  virtual GridType GetGridType() const override;
```
Comment (gaschler) [inline cartographer/mapping/2d/probability_grid.h]: drop virtual in derived

File: cartographer/mapping/2d/tsdf_2d.h [hunk 10]
```diff
@@ -37,6 +37,7 @@ class TSDF2D : public Grid2D {
 
   void SetCell(const Eigen::Array2i& cell_index, const float tsd,
                const float weight);
+  virtual GridType GetGridType() const override;
```
Comment (gaschler) [inline cartographer/mapping/2d/tsdf_2d.h]: drop virtual

**Comment by kdaun** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user kdaun. Merge queue now has a length of 2.

**Pull request merged**